File: tools/ticker_analytics_lookup_tool.py
from crewai.tools import BaseTool
from typing import Type
from pydantic import BaseModel, Field
from elasticsearch import Elasticsearch
import os

class TickerAnalyticsLookupToolInput(BaseModel):
    """Input schema for TickerAnalyticsLookupTool."""
    ticker: str = Field(..., description="Ticker symbol of the stock.")

class TickerAnalyticsLookupTool(BaseTool):
    name: str = "TickerAnalyticsLookupTool"
    description: str = (
        "Lookup ticker analytics from elasticsearch to obtain technical data, fundamental data and news summary."
    )
    args_schema: Type[BaseModel] = TickerAnalyticsLookupToolInput

    def _run(self, ticker: str) -> str:

        ELASTIC_SEARCH_URL = os.getenv('ELASTIC_SEARCH_URL')
        ES_API_KEY = os.getenv('ES_API_KEY')
        elastic_client = Elasticsearch(hosts=ELASTIC_SEARCH_URL, api_key=ES_API_KEY)
        body = {
            "query": {
                "match": {
                    "ticker": f'{ticker}'
                }
            },
            "size": 1,
            "sort": [
                {
                    "timestamp": {
                        "order": "desc"
                    }
                }
            ]
        }
        results = elastic_client.search(body=body, index="ticker_analytics")
        ticker_data = [hit['_source'] for hit in results['hits']['hits']]
        print(f"Found {len(ticker_data)} ticker_analytics documents in elastic search.")
        return ticker_data

File: mastra/src/mastra/tools/ticker-analytics-lookup-tool.ts
import { createTool } from "@mastra/core/tools";
import { z } from "zod";
import { config } from "dotenv";
import { Client } from '@elastic/elasticsearch';

config();

const inputSchema = z.object({
  ticker: z
    .string()
    .describe("Ticker symbol to search for."),
  size: z
    .number()
    .int()
    .min(1)
    .max(30)
    .default(10)
    .describe("Number of documents to retrieve (1-30), default 10."),
});

const outputSchema = z.object({
  success: z.boolean(),
  message: z.string(),
  data: z.array(z.any()).optional(),
});

// Initialize Elasticsearch client
const client = new Client({ 
  node: process.env.ELASTIC_SEARCH_URL!,
  auth: {
    apiKey: process.env.ES_API_KEY!
  }
});

const executeTickerAnalyticsLookupTool = async (input: z.infer<typeof inputSchema>): Promise<z.infer<typeof outputSchema>> => {
  try {
    if (!process.env.ELASTIC_SEARCH_URL) {
      throw new Error("Elasticsearch URL is not set. Ensure ELASTIC_SEARCH_URL is defined in the .env file.");
    }

    if (!process.env.ES_API_KEY) {
      throw new Error("Elasticsearch API key is not set. Ensure ES_API_KEY is defined in the .env file.");
    }

    // Log input for debugging
    console.log("Input received:", input);

    // Extract parameters from input.context
    const { ticker, size } = input.context;

    // Validate ticker explicitly
    if (!ticker || typeof ticker !== 'string' || ticker.trim() === '') {
      throw new Error("The 'ticker' parameter is required and must be a non-empty string.");
    }

    // Validate input using zod schema
    const validatedInput = inputSchema.parse({ ticker, size });

    // Log validated input
    console.log("Validated input:", validatedInput);

    // Prepare search template parameters
    const templateParams: Record<string, any> = {
      ticker: validatedInput.ticker,
      size: validatedInput.size,
    };

    // Log template parameters
    console.log("Template parameters:", templateParams);

    // Use search template
    const response = await client.searchTemplate({
      index: 'ticker_analytics',
      id: 'ticker_analytics_lookup_template',
      params: templateParams
    });

    return {
      success: true,
      message: `Retrieved ${response.hits.hits.length} documents from Elasticsearch for ticker: ${validatedInput.ticker}.`,
      data: response.hits.hits,
    };
  } catch (error: any) {
    return {
      success: false,
      message: `Error querying Elasticsearch: ${error.message}`,
    };
  }
};

export const tickerAnalyticsLookupTool = createTool({
  id: "ticker-analytics-lookup-tool",
  description: "Search ticker analytics data from Elasticsearch using the ticker_analytics_lookup_template. Requires a ticker symbol.",
  inputSchema,
  outputSchema,
  execute: async (input: any) => {
    // Pass input directly to executeTickerAnalyticsLookupTool, letting zod handle defaults
    return executeTickerAnalyticsLookupTool(input);
  },
});
